File: module3/func-oop/animals_classes.js
/* 
Animal - Bird - Parrot
    |
    --- Cat - House Cat
        |
        Tiger
*/

// 2 essentiak keywords for OOP: 
// extends (to inherit from an existing class) 
// and super (to "borrow" functionality from a super-class)

// ! creating animal classes

class Animal {
    constructor(color = 'yellow', energy = 100) {
        this.color = color;
        this.energy = energy;
    }
    isActive() {
        if (this.energy > 0) {
            this.energy -= 20;
            console.log("Energy is decreasing, currently at: ", this.energy);
        } else if (this.energy == 0) {
            this.sleep();
        }
    }
    sleep(){
        this.energy += 20;
        console.log('Energy is increasing, currently at:', this.energy);
    }
    getColor(){
        console.log("Color: ", this.color);
    }
}

/* If I didn't use the super keyword in our sub-classes, I'd get the following error: 
Uncaught ReferenceError: Must call super constructor in derived class 
before accessing 'this' or returning from derived constructor.
*/

class Cat extends Animal {
    constructor(sound = 'purr', canJumpHigh = true, canClimbTrees = true, color, energy)  {
        super(color, energy);
        this.sound = sound;
        this.canJumpHigh = canJumpHigh;
        this.canClimbTrees = canClimbTrees;
    }
    makeSound(){
        console.log(this.sound);
    }
}

class HouseCat extends Cat {
    constructor(color, energy, sound, canJumpHigh, canClimbTrees, houseCatSound = 'meow') {
        super(color, energy, sound,canJumpHigh, canClimbTrees);
        this.houseCatSound = houseCatSound;
    }
    // override method
    makeSound(option){
        if (option) {
            super.makeSound();
        }
        console.log(this.houseCatSound);
    }
}

class Tiger extends Cat {
    constructor(tigerSound='Roar!', color, energy, sound, canJumpHigh, canClimbTrees, houseCatSound){
        super(color, energy, sound, canJumpHigh, canClimbTrees, houseCatSound);
        this.tigerSound = tigerSound;
    }
    makeSound(option) {
        if (option) {
            super.makeSound();
        }    
        console.log(this.tigerSound);
    }
}

class Bird extends Animal {
    constructor(sound = 'chirp', canFly = true, color, energy) {
        super(color, energy);
        this.sound = sound;
        this.canFly = canFly;
    }
    makeSound() {
        console.log(this.sound);
    }
}

class Parrot extends Bird {
    constructor(canTalk = false, sound, canFly, color, energy) {
        super(sound, canFly, color, energy);
        this.canTalk = canTalk;
    }
    makeSound(option) {
        if (option) {
            super.makeSound();
        }
        if (this.canTalk) {
            console.log("I'm a talking parrot!");
        }
    }
}

// ! now we can build various animal objects
// two parrots: one that can talk, and the other that can't

var polly = new Parrot(true);
var fiji = new Parrot(false);

console.log("Polly says: ")
polly.makeSound();

console.log("Fiji says: ")
fiji.makeSound(true);

console.log(`Polly color is ${polly.color} and energy is ${polly.energy}`);

polly.isActive();

var penguin = new Bird("shriek", false, 'black and white', 200);
console.log('Penguin: ', penguin)

var leo = new HouseCat();
console.log('Leo says: ')
// leo, no purring please:
leo.makeSound(false);
// leo, both purr and meow now:
leo.makeSound(true);

File: module3/func-oop/animals_classes.py
''' 
Animal - Bird - Parrot
    |
    --- Cat - House Cat
        |
        Tiger
'''

# rewrite js example with animals classes in python 

class Animal:
    def __init__(self, color = 'yellow', energy = 100):
        self.color = color
        self.energy = energy
    
    def is_active(self):
        if self.energy > 0:
            self.energy -= 20
            print("Energy is decreasing, currently at: ", self.energy)
        else:
            self.sleep()
    
    def sleep(self):
        self.energy += 20
        print('Energy is increasing, currently at:', self.energy)

    def get_color(self):
        print('Color: ', self.color)    

class Cat(Animal):
    def __init__(self, sound = 'purr', can_jump_high = True, can_climb_trees = True):
        super().__init__()
        self.can_jump_high = can_jump_high
        self.can_climb_trees = can_climb_trees
        self.sound = sound
    
    def make_sound(self):
        print(self.sound)

class HouseCat(Cat):
    def __init__(self, house_cat_sound = 'meow'):
        super().__init__()
        self.house_cat_sound = house_cat_sound
    
    def make_sound(self, option):
        if option:
            super().make_sound()
        print(self.house_cat_sound)

class Tiger(Cat):
    def __init__(self, tiger_sound = 'roar!'):
        super().__init__()
        self.tiger_sound = tiger_sound
    
    def make_sound(self, option):
        if option:
            super().make_sound()
        print(self.tiger_sound)

class Bird(Animal):
    def __init__(self, can_fly = True, sound='chirp'):
        super().__init__()
        self.sound = sound
        self.can_fly = can_fly
    
    def make_sound(self):
        print(self.sound)

class Parrot(Bird):
    def __init__(self, can_talk = True):
        super().__init__()
        self.can_talk = can_talk
    
    def make_sound(self, option):
        if option:
            super().make_sound()
        if self.can_talk:
            print("I'm talking parrot")

polly = Parrot(can_talk=True)
fiji = Parrot(can_talk=False)

print("Parrot Polly: ")
polly.color = 'red'
polly.get_color()
polly.make_sound(True)


print("Parrot Fiji: ")
fiji.make_sound(False)

penguin = Bird(False, 'shriek')
penguin.color = 'black and white'
penguin.energy = 200
print(penguin)